front/lib/api/assistant/agent.ts

```diff
@@ -270,14 +310,414 @@ export async function getNextAction(
-      : configuration.actions.find((a) => a.name === output.name);
+      ? agentConfiguration.generation
+      : agentConfiguration.actions.find((a) => a.name === output.name);
+
+  const spec = specifications.find((s) => s.name === output.name) ?? null;
+
   if (!action) {
     return new Err(new Error(`Action ${output.name} not found`));
   }
 
   return new Ok({
     action,
     inputs: output.arguments,
+    specification: spec,
   });
 }
+
+export async function* runMultiActionsAgent(
+  auth: Authenticator,
+  configuration: AgentConfigurationType,
+  conversation: ConversationType,
+  userMessage: UserMessageType,
+  agentMessage: AgentMessageType
+): AsyncGenerator<
+  | AgentErrorEvent
+  | AgentActionEvent
+  | AgentActionSuccessEvent
+  | GenerationTokensEvent
+  | AgentGenerationSuccessEvent
+  | AgentGenerationCancelledEvent
+  | AgentMessageSuccessEvent,
+  void
+> {
+  const now = Date.now();
+
+  for (let i = 0; i < configuration.maxToolsUsePerRun; i++) {
+    const localLogger = logger.child({
+      workspaceId: conversation.owner.sId,
+      conversationId: conversation.sId,
+      multiActionLoopIteration: i,
+    });
+
+    localLogger.info("Starting multi-action loop iteration");
+
+    const forcedAction = configuration.actions.find(
+      (a) => a.forceUseAtIteration === i
+    );
+    const actions = forcedAction ? [forcedAction] : configuration.actions;
+
+    const actionToRun = await getNextAction(auth, {
+      agentConfiguration: configuration,
+      conversation,
+      userMessage,
+      availableActions: actions,
+      isGenerationAllowed: !forcedAction,
+    });
+
+    if (actionToRun.isErr()) {
+      localLogger.error(
+        {
+          elapsedTime: Date.now() - now,
+          error: actionToRun.error,
+        },
+        "Error getting next action"
+      );
+      yield {
+        type: "agent_error",
+        created: now,
+        configurationId: configuration.sId,
+        messageId: agentMessage.sId,
+        error: {
+          code: "parameters_generation_error",
+          message: `Error getting next action: ${actionToRun.error.message}`,
+        },
+      };
+      return;
+    }
+
+    localLogger.info(
+      {
+        elapsed: Date.now() - now,
+      },
+      "[ASSISTANT_TRACE] Action inputs generation"
+    );
+
+    const { action, inputs, specification } = actionToRun.value;
+
+    if (isAgentConfiguration(action)) {
+      const eventStream = runAction(auth, {
+        configuration: configuration,
+        actionConfiguration: action,
+        conversation,
+        userMessage,
+        agentMessage,
+        inputs,
+        specification,
+      });
+      for await (const event of eventStream) {
+        yield event;
+      }
+    } else {
+      // TODO(@fontanierh): remove this once generation is part of actions.
+      // This is just to assert that we cover all action types.
+      ((g: AgentGenerationConfigurationType) => {
+        void g;
+      })(action);
```

Comment:
Why do we need those lines of code for?

Reply:
This allows to use the typs system to ensure we didn't forget to handle an action type. Similar to an assertNever (but here it's "assert action is indeed a generation config"). As explained in the comment above